horovod/common/operations.cc

```diff
@@ -843,26 +897,80 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
                                      cudaMemcpyDeviceToDevice, stream))
           offset += it->tensor->size();
         }
+        buffer_len = (size_t)offset;
         if (timeline.Initialized()) {
-          RECORD_EVENT(entries, after_memcpy_in_event, stream)
+          RECORD_EVENT(entries, event_queue, "MEMCPY_IN_FUSION_BUFFER", stream)
         }
 
         // Perform the reduction on the fusion buffer.
-        int64_t num_elements = 0;
         for (auto it = entries.begin(); it != entries.end(); it++) {
           num_elements += it->tensor->shape().num_elements();
         }
+      } else {
+        buffer_data = (void*)first_entry.output->data();
+        num_elements = first_entry.tensor->shape().num_elements();
+        buffer_len = (size_t)first_entry.output->size();
+      }
+
+      void* host_buffer = nullptr;
+      if (horovod_global.hierarchical_allreduce) {
+        NCCL_CHECK(entries, "ncclReduce",
+                   ncclReduce(buffer_data, buffer_data, (size_t)num_elements,
+                              GetNCCLDataType(first_entry.tensor), ncclSum, 0,
+                              nccl_comm, stream))
+        if (timeline.Initialized()) {
+          RECORD_EVENT(entries, event_queue, "NCCL_REDUCE", stream)
```

Comment:
Generally, for all string constant, I think it would be more maintainable and safer to have them be defined with global variables.